libraries/eosiolib/contracts/eosio/key_value_singleton.hpp

```diff
@@ -0,0 +1,164 @@
+
+         __attribute__((eosio_wasm_import))
+         int64_t kv_set(uint64_t db, uint64_t contract, const char* key, uint32_t key_size, const char* value, uint32_t value_size);
+
+         __attribute__((eosio_wasm_import))
+         bool kv_get(uint64_t db, uint64_t contract, const char* key, uint32_t key_size, uint32_t& value_size);
+
+         __attribute__((eosio_wasm_import))
+         uint32_t kv_get_data(uint64_t db, uint32_t offset, char* data, uint32_t data_size);
+      }
+   }
+
+   template <typename T, eosio::name::raw SingletonName, eosio::name::raw DbName = "eosio.kvram"_n>
+   class kv_singleton {
+
+      struct state {
+         T value;
+         key_type raw_original;
+         bool is_dirty = false;
+         bool is_cached = false;
+      };
+
+   public:
+      kv_singleton(eosio::name contract_name) : contract_name{contract_name} {
+         key = make_prefix(singleton_name);
+      }
+
+      ~kv_singleton() {
+         if (get_state().is_dirty) {
+            store();
+         }
+      }
+
+      const T& get() {
+         auto& ste = get_state();
+         if (!ste.is_cached) {
+            uint32_t copy_size;
+            uint32_t value_size;
+
+            auto success = internal_use_do_not_use::kv_get(db_name, contract_name.value, key.data(), key.size(), value_size);
+
+            eosio::check(success, "tried to get a singleton that does not exist");
+
+            void* val_buffer = value_size > detail::max_stack_buffer_size ? malloc(value_size) : alloca(value_size);
+            copy_size = internal_use_do_not_use::kv_get_data(db_name, 0, (char*)val_buffer, value_size);
+
+            T val;
+            deserialize(val, val_buffer, copy_size);
+
+            ste.value = val;
+            ste.is_cached = true;
+
+            ste.raw_original.resize(copy_size);
+            memcpy(ste.raw_original.data(), val_buffer, copy_size);
+
+            if (value_size > detail::max_stack_buffer_size) {
+               free(val_buffer);
+            }
+         }
+
+         return get_state().value;
+      }
+
+      void set(const T& val) {
+         auto& ste = get_state();
+         ste.value = val;
+         ste.is_dirty = true;
+         ste.is_cached = true;
+      }
+
+      void erase() {
+         internal_use_do_not_use::kv_erase(db_name, contract_name.value, key.data(), key.size());
+         auto& ste = get_state();
+         ste.is_cached = false;
+         ste.is_dirty = false;
+      }
+
+      void store() {
+         auto& ste = get_state();
+         if (ste.is_dirty) {
+            size_t data_size = get_size(ste.value);
+            void* data_buffer = data_size > detail::max_stack_buffer_size ? malloc(data_size) : alloca(data_size);
+
+            serialize(ste.value, data_buffer, data_size);
+
+            if (ste.raw_original.size() != data_size || memcmp(ste.raw_original.data(), data_buffer, data_size) != 0) {
```

Comment:
Is it a problem that bypasses access control and allows any contract to call set as long as it sets it to its original value?

Reply:
A stranger case is setting it to a new value, and then setting it back to the old value.

Reply:
What do you mean by bypasses access control?

Reply:
I mean the restriction that a contract can only write to its own kv store.

Reply:
Doesn't the `contract_name.value` in `internal_use_do_not_use::kv_set(db_name, contract_name.value, key.data(), key.size(), (const char*)data_buffer, data_size);` handle that restriction? I thought the intrinsic would enforce that the contract passed in is the same as the contract that its running in.

Reply:
It does--but only if kv_set is actually called.  To be clear, this doesn't allow the contract to cause problems for another contract.  It just means that the caching behavior is observable and code that is probably incorrect will sometimes "work."

Reply:
Ah I understand now. I've considered removing this specific check, and just always call `kv_set` if `is_dirty` to simplify the code.